vcd/resource_vcd_nsxt_edgegateway_bgp_configuration.go

```diff
@@ -0,0 +1,248 @@
+	vcdClient := meta.(*VCDClient)
+
+	// Handling locks on for BGP configuration is conditional. There are two scenarios:
+	// * When the parent Edge Gateway is in a VDC - a lock on parent Edge Gateway must be acquired
+	// * When the parent Edge Gateway is in a VDC Group - a lock on parent VDC Group must be acquired
+	// To find out parent lock object, Edge Gateway must be looked up and its OwnerRef must be checked
+	// Note. It is not safe to do multiple locks in the same resource as it can result in a deadlock
+	parentEdgeGatewayOwnerId, _, err := getParentEdgeGatewayOwnerId(vcdClient, d)
+	if err != nil {
+		return diag.Errorf("[route advertisement delete] error finding parent Edge Gateway: %s", err)
+	}
+
+	if govcd.OwnerIsVdcGroup(parentEdgeGatewayOwnerId) {
+		vcdClient.lockById(parentEdgeGatewayOwnerId)
+		defer vcdClient.unlockById(parentEdgeGatewayOwnerId)
+	} else {
+		vcdClient.lockParentEdgeGtw(d)
+		defer vcdClient.unLockParentEdgeGtw(d)
+	}
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+
+	if err != nil {
+		return diag.Errorf("error retrieving NSX-T Edge Gateway: %s", err)
+	}
+
+	err = nsxtEdge.DisableBgpConfiguration()
+	if err != nil {
+		return diag.Errorf("error disableing NSX-T Edge Gateway BGP Configuration: %s", err)
+	}
+
+	return nil
+}
+
+func resourceVcdEdgeBgpConfigImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	log.Printf("[TRACE] NSX-T Edge Gateway BGP Configuration import initiated")
+
+	resourceURI := strings.Split(d.Id(), ImportSeparator)
+	if len(resourceURI) != 3 {
+		return nil, fmt.Errorf("resource name must be specified as org-name.vdc-or-vdc-group-name.nsxt-edge-gw-name")
+	}
+	orgName, vdcOrVdcGroupName, edgeName := resourceURI[0], resourceURI[1], resourceURI[2]
+
+	vcdClient := meta.(*VCDClient)
+	vdcOrVdcGroup, err := lookupVdcOrVdcGroup(vcdClient, orgName, vdcOrVdcGroupName)
+	if err != nil {
+		return nil, err
+	}
+
+	edge, err := vdcOrVdcGroup.GetNsxtEdgeGatewayByName(edgeName)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve NSX-T edge gateway with ID '%s': %s", d.Id(), err)
+	}
+
+	dSet(d, "org", orgName)
+	dSet(d, "edge_gateway_id", edge.EdgeGateway.ID)
+	d.SetId(edge.EdgeGateway.ID)
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func getEdgeBgpConfigType(d *schema.ResourceData) *types.EdgeBgpConfig {
+	bgpConfig := &types.EdgeBgpConfig{
+		Enabled: d.Get("enabled").(bool),
+		Ecmp:    d.Get("ecmp_enabled").(bool),
+		// Version is required, but it is automatically handled by Go Cloud Director SDK in function
+		// UpdateBgpConfiguration
+	}
+
+	// bgpConfig.GracefulRestart can only be specified when Edge Gateway is backed by Tier-0
+	// gateway, not a VRF. For that reason types.EdgeBgpGracefulRestartConfig must only be sent if
+	// the user configured it.
```

Comment:
if we don't check T-0 isn't VRF -> if API error is good enough for the user to understand?

Reply:
Yep - the error is human-readable in that case.